### ISSUE TITLE
Unable to use delete key or backspace

### ISSUE THREAD

**Issue by jhilker1** [jwlodek/py_cui#74]
**Describe the bug**
Hitting the backspace or delete keys appends '^?' to the text I enter instead of deleting it.

**To Reproduce**
1. Write some text in any widget
2. Press delete or backspace

**Expected behavior**
If I hit delete, the character the cursor is on should be deleted, but if I hit backspace, the previous character should be deleted.

**Screenshots**
Will update tomorrow with screenshots

**Environment*:*
 - OS: Ubuntu 18.04
 - Terminal: Alacritty
 - Version: 0.1.2

**Comment by jwlodek** [comment]
I will try and reproduce this tomorrow. Do you have this bug only in Alacritty, or in other terminal emulators on your system as well? There have been some inconsistencies in behavior of `py_cui` in different terminals, though mostly on windows, I haven't seen something like this on windows.

**Comment by jhilker1** [comment]
I just tested in Gnome Terminal and had the same issue.

**Comment by jwlodek** [comment]
Which widget/popup specifically is seeing this behavior?

**Comment by jhilker1** [comment]
It is on the text box widgets and pop-ups.

**Comment by JeremyBYU** [comment]
I have the same exact issue as well. Ubuntu 18.04 gnome terminal.

**Comment by blackCauldron7** [comment]
Same issue although Jeremy's solution fixes it.

**Comment by jwlodek** [comment]
This is odd, because I tried to reproduce the issue on my Ubuntu 18.04 machine with gnome-terminal, but I couldn't get that kind of behavior. I wonder what this could possibly be caused by?

**Comment by blackCauldron7** [comment]
I was using alacritty.

**Comment by jwlodek** [comment]
I haven't tried alacritty, but @jhilker1 mentioned he tried gnome-terminal and saw the same behavior. It seems that the `curses` pre-defined backspace key code is not robust, we may need to consider some logic for checking if a key is one of the several possible backspace codes

**Comment by JeremyBYU** [comment]
Yeah, I think you may need to check for a few possible values:

https://stackoverflow.com/a/54303430

**Comment by lenormf** [comment]
I could reproduce the issue on Kitty.

**Comment by voryzen** [comment]
+1 this issue, using konsole, zsh

**Comment by jwlodek** [comment]
@voryzen could you run the following curses program on your system and let me know what value of keypress you get when pressing backspace/delete?

```Python
import sys,os
import curses

def draw_menu(stdscr):
    k = 0
    cursor_x = 0
    cursor_y = 0

    # Clear and refresh the screen for a blank canvas
    stdscr.clear()
    stdscr.refresh()

    # Start colors in curses
    curses.start_color()
    curses.init_pair(1, curses.COLOR_CYAN, curses.COLOR_BLACK)
    curses.init_pair(2, curses.COLOR_RED, curses.COLOR_BLACK)
    curses.init_pair(3, curses.COLOR_BLACK, curses.COLOR_WHITE)

    # Loop where k is the last character pressed
    while (k != ord('q')):

        # Initialization
        stdscr.clear()
        height, width = stdscr.getmaxyx()

        if k == curses.KEY_DOWN:
            cursor_y = cursor_y + 1
        elif k == curses.KEY_UP:
            cursor_y = cursor_y - 1
        elif k == curses.KEY_RIGHT:
            cursor_x = cursor_x + 1
        elif k == curses.KEY_LEFT:
            cursor_x = cursor_x - 1

        cursor_x = max(0, cursor_x)
        cursor_x = min(width-1, cursor_x)

        cursor_y = max(0, cursor_y)
        cursor_y = min(height-1, cursor_y)

        # Declaration of strings
        title = "Curses example"[:width-1]
        subtitle = "Written by Clay McLeod"[:width-1]
        keystr = "Last key pressed: {}".format(k)[:width-1]
        statusbarstr = "Press 'q' to exit | STATUS BAR | Pos: {}, {}".format(cursor_x, cursor_y)
        if k == 0:
            keystr = "No key press detected..."[:width-1]

        # Centering calculations
        start_x_title = int((width // 2) - (len(title) // 2) - len(title) % 2)
        start_x_subtitle = int((width // 2) - (len(subtitle) // 2) - len(subtitle) % 2)
        start_x_keystr = int((width // 2) - (len(keystr) // 2) - len(keystr) % 2)
        start_y = int((height // 2) - 2)

        # Rendering some text
        whstr = "Width: {}, Height: {}".format(width, height)
        stdscr.addstr(0, 0, whstr, curses.color_pair(1))

        # Render status bar
        stdscr.attron(curses.color_pair(3))
        stdscr.addstr(height-1, 0, statusbarstr)
        stdscr.addstr(height-1, len(statusbarstr), " " * (width - len(statusbarstr) - 1))
        stdscr.attroff(curses.color_pair(3))

        # Turning on attributes for title
        stdscr.attron(curses.color_pair(2))
        stdscr.attron(curses.A_BOLD)

        # Rendering title
        stdscr.addstr(start_y, start_x_title, title)

        # Turning off attributes for title
        stdscr.attroff(curses.color_pair(2))
        stdscr.attroff(curses.A_BOLD)

        # Print rest of text
        stdscr.addstr(start_y + 1, start_x_subtitle, subtitle)
        stdscr.addstr(start_y + 3, (width // 2) - 2, '-' * 4)
        stdscr.addstr(start_y + 5, start_x_keystr, keystr)
        stdscr.move(cursor_y, cursor_x)

        # Refresh the screen
        stdscr.refresh()

        # Wait for next input
        k = stdscr.getch()

def main():
    curses.wrapper(draw_menu)

if __name__ == "__main__":
    main()
```

**Comment by voryzen** [comment]
Hi J,

127

I used both konsole and gnome-terminal

**Comment by voryzen** [comment]
@jwlodek if it helps at all, these are the changes that I've made to my py_cui install
````
# Pressing backspace returns 8 on windows?
if platform == 'win32':
    KEY_BACKSPACE   = 8
# Adds support for 'delete/backspace' key on OSX
elif platform == 'darwin':
    KEY_BACKSPACE   = 127
elif platform == 'linux':
	KEY_BACKSPACE	= 127
else:
    KEY_BACKSPACE   = curses.KEY_BACKSPACE
````

**Comment by voryzen** [comment]
HI @PabloLec,
After installing your backspace fix.

```
  File "/home/voryzen/Downloads/py_cui/py_cui/widgets.py", line 102, in add_key_command
    self._keybindings[key] = command
AttributeError: 'ScrollMenu' object has no attribute '_keybindings'
```

**Comment by PabloLec** [comment]
Hi!

Yes this is fixed by #159 .
